### PR TITLE
merge bitcoin#15931, #16839, #17192,  #17407, #18037, #17997, partial #15639, #17989: deglobalisation and mining rpc backports

### PR DESCRIPTION
**Pull request by kittywhiskers**
## Checklist

**Part 1 (this PR)**
- [x] https://github.com/bitcoin/bitcoin/pull/16839 (relies on https://github.com/dashpay/dash/pull/4443, which is what this PR was originally supposed to address)
- [x] https://github.com/bitcoin/bitcoin/pull/17997

**Part _n_**
- [ ] https://github.com/bitcoin/bitcoin/pull/18038
- [ ] https://github.com/bitcoin/bitcoin/pull/18740
- [ ] https://github.com/bitcoin/bitcoin/pull/19096
- [ ] https://github.com/bitcoin/bitcoin/pull/19413
- [ ] https://github.com/bitcoin/bitcoin/pull/19779

## Dependencies

- https://github.com/bitcoin/bitcoin/pull/15639
- https://github.com/bitcoin/bitcoin/pull/15931
- https://github.com/bitcoin/bitcoin/pull/16839
- https://github.com/bitcoin/bitcoin/pull/17192
- https://github.com/bitcoin/bitcoin/pull/17407
- https://github.com/bitcoin/bitcoin/pull/17989
- https://github.com/bitcoin/bitcoin/pull/17997
- https://github.com/bitcoin/bitcoin/pull/18037

### CODE REVIEW

File: src/interfaces/chain.cpp [hunk 1]
```diff
@@ -187,11 +187,11 @@ class NotificationsHandlerImpl : public Handler, CValidationInterface
         const CBlockIndex* index,
         const std::vector<CTransactionRef>& tx_conflicted) override
     {
-        m_notifications->BlockConnected(*block, tx_conflicted);
```
Comment (PastaPastaPasta) [inline src/interfaces/chain.cpp:190]: should drop `getBlockDepth` no? https://github.com/bitcoin/bitcoin/pull/15931/files#diff-efaf73429006d059e7fffab1bf2879769c90187f583d4857b3275b7bf79368d1L61-L66

Comment (PastaPastaPasta) [reply src/interfaces/chain.cpp:190]: Also, is there a reason we don't currently have waitForNotificationsIfNewBlocksConnected? https://github.com/bitcoin/bitcoin/pull/15931/files#diff-efaf73429006d059e7fffab1bf2879769c90187f583d4857b3275b7bf79368d1L356-L362

File: src/interfaces/chain.h [hunk 2]
```diff
@@ -235,8 +235,8 @@ class Chain
         virtual ~Notifications() {}
         virtual void TransactionAddedToMempool(const CTransactionRef& tx, int64_t nAcceptTime) {}
         virtual void TransactionRemovedFromMempool(const CTransactionRef& ptx, MemPoolRemovalReason reason) {}
-        virtual void BlockConnected(const CBlock& block, const std::vector<CTransactionRef>& tx_conflicted) {}
-        virtual void BlockDisconnected(const CBlock& block) {}
+        virtual void BlockConnected(const CBlock& block, const std::vector<CTransactionRef>& tx_conflicted, int height) {}
+        virtual void BlockDisconnected(const CBlock& block, int height) {}
```
Comment (PastaPastaPasta) [inline src/interfaces/chain.h:241]: same missing changes mentioned above

File: src/qt/test/wallettests.cpp [hunk 3]
```diff
@@ -113,6 +113,7 @@ void TestGUI()
         LOCK(wallet->cs_wallet);
         wallet->SetAddressBook(test.coinbaseKey.GetPubKey().GetID(), "", "receive");
         wallet->AddKeyPubKey(test.coinbaseKey, test.coinbaseKey.GetPubKey());
+        wallet->SetLastBlockProcessed(105, ::ChainActive().Tip()->GetBlockHash());
```
Comment (PastaPastaPasta) [inline src/qt/test/wallettests.cpp:116]: missing changes in interfaces/wallet.cpp (didn't review all of 15931, moving on for now due to missing changes)

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
needs rebase + pls see https://github.com/UdjinM6/dash/commits/pr4357

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
LGTM but gitlab action no longer works for this PR for some reason :/ should replace this PR with a new one

**Comment by kittywhiskers** [comment]
Superseded by #4756

**Pull request closed without merging**